Fix and reapply IR PGO support for Flang

fanju110 opened the pull request:
This PR resubmits the changes from #136098, which was previously reverted due to a build failure during the linking stage:

```
undefined reference to `llvm::DebugInfoCorrelate'  
undefined reference to `llvm::ProfileCorrelate'
```

The root cause was that `llvm/lib/Frontend/Driver/CodeGenOptions.cpp` references symbols from the `Instrumentation` component, but the `LINK_COMPONENTS` in the `llvm/lib/Frontend/CMakeLists.txt` for `LLVMFrontendDriver` did not include it. As a result, linking failed in configurations where these components were not transitively linked.

### Fix:

This updated patch explicitly adds `Instrumentation` to `LINK_COMPONENTS` in the relevant `llvm/lib/Frontend/CMakeLists.txt` file to ensure the required symbols are properly resolved.

fanju110 committed:
Fix and reapply IR PGO support for Flang

Fix and reapply IR PGO support for Flang

Co-Authored-By: ict-ql <168183727+ict-ql@users.noreply.github.com>
Co-Authored-By: Chyaka <52224511+liliumshade@users.noreply.github.com>

fanju110 commented:
@tarunprabhu Hi, I have fixed the previous problem. The exact reason is in the description of this PR. We can do all the checks manually and pass them all before merging. Thanks.

fanju110 committed:
Merge branch 'main' into OpenXiangShan/reapply-flang-pgo

fanju110 commented:
> Thanks for the fix and the explanation. I will try this PR out, but it might take me a day or two. Do you have merge access? If not, I will merge it if everything passes for me.

I don’t have merge access.  I’ve run `check-clang`, `check-flang`, and `check-llvm` locally, and also re-tested the previously failing build configuration. They look good on my end. If everything also looks good on your side,Please help merge it when convenient.Thank you very much.

tarunprabhu reviewed (changes requested):
Apologies for the delay. I had to proritize other things and I lost track of this. 

I checked with `-DLLVM_LINK_LLVM_DYLIB=OFF`, `-DLLVM_LINK_LLVM_DYLIB=ON` and `-DLLVM_BUILD_SHARED_LIBS=ON` and all three built and passed checks. 

I will merge this after you address the issue with the `switch` statement

fanju110 committed:
Update clang/lib/Basic/ProfileList.cpp

Update the  label  to match the expected enum type and resolve the compiler warning about comparing different enumeration types.

Co-authored-by: Tarun Prabhu <tarunprabhu@gmail.com>

fanju110 commented on clang/lib/Basic/ProfileList.cpp:
```diff
@@ -69,24 +69,24 @@ ProfileList::ProfileList(ArrayRef<std::string> Paths, SourceManager &SM)
-  case CodeGenOptions::ProfileCSIRInstr:
+  case llvm::driver::ProfileInstrKind::ProfileCSIRInstr:
     return "csllvm";
   case CodeGenOptions::ProfileIRSampleColdCov:
```
Thank you very much, I've changed it here

fanju110 replied:
This warning is probably due to a duplicate definition of `ProfileInstrKind`. Because the flang IR PGO does not use `ProfileIRSampleColdCov`, the original code left it in `ProfileInstrKind `in `clang`. Today I move `ProfileIRSampleColdCov `to `llvm `and removed the definition in `clang`.

fanju110 replied:
> I get a warning about comparing different enums. I assume that the suggestion below is what this is supposed to be.
> 
> ```suggestion
>   case llvm::driver::ProfileInstrKind::ProfileIRSampleColdCov:
> ```

Following your test, I also test -DLLVM_LINK_LLVM_DYLIB=OFF, -DLLVM_LINK_LLVM_DYLIB=ON and -DLLVM_BUILD_SHARED_LIBS=ON at my host,which all three built successfully and passed the test

fanju110 committed:
Merge branch 'main' into OpenXiangShan/reapply-flang-pgo

fanju110 committed:
move ProfileIRSampleColdCov from clang to llvm and Remove ProfileInstrKind definition in clang

tarunprabhu reviewed (approved):
Thanks. I'll merge this tomorrow morning (my time), so I can keep an eye on the buildbots, just in case.

fanju110 commented:
> Thanks. I'll merge this tomorrow morning (my time), so I can keep an eye on the buildbots, just in case.
Thank you for your help.There is a test-case test that fails, but I don't reproduce the error when I check-flang locally, is this problem normal? Could it be related to my code?

fanju110 commented:
> Thanks. I'll merge this tomorrow morning (my time), so I can keep an eye on the buildbots, just in case.


Thanks. I'll merge this tomorrow morning (my time), so I can keep an eye on the buildbots, just in case.
Thank you for your help.There is a test-case test that fails, but I don't reproduce the error when I check-flang locally, is this problem normal? Could it be related to my code?

tarunprabhu commented:
There is a similar build failure [here](https://github.com/llvm/llvm-project/pull/143603) that seems to be even older. I don't think these are related.

kiranchandramohan commented:
The test added in this patch is failing since it cannot find llvm-profdata. 
https://github.com/llvm/llvm-project/actions/runs/15643574721/job/44076443329?pr=144143

Attempting a fix https://github.com/llvm/llvm-project/pull/144325

tarunprabhu commented:
Thanks for the fix @kiranchandramohan. I didn't spot the issue with `llvm-profdata`
